File: backend/src/models/Supplier.js
const mongoose = require('mongoose');

const supplierSchema = new mongoose.Schema(
  {
    code: {
      type: String,
      unique: true,
      trim: true,
      uppercase: true,
    },
    name: {
      type: String,
      required: [true, 'Vui lòng nhập tên nhà cung cấp'],
      trim: true,
    },
    contactPerson: {
      type: String,
      trim: true,
    },
    phone: {
      type: String,
      required: [true, 'Vui lòng nhập số điện thoại'],
      trim: true,
      match: [/^[0-9]{10,11}$/, 'Số điện thoại không hợp lệ'],
    },
    email: {
      type: String,
      trim: true,
      lowercase: true,
      match: [
        /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/,
        'Email không hợp lệ',
      ],
    },
    address: {
      street: String,
      district: String,
      city: String,
      country: {
        type: String,
        default: 'Việt Nam',
      },
    },
    taxCode: {
      type: String,
      trim: true,
    },
    bankAccount: {
      bankName: String,
      accountNumber: String,
      accountName: String,
    },
    
    isActive: {
      type: Boolean,
      default: true,
    },
    createdBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
  },
  {
    timestamps: true,
  }
);

// Middleware: Tạo mã tự động nếu chưa có
supplierSchema.pre('save', async function(next) {
  if (!this.code && this.isNew) {
    try {
      // Generate code SUP01, SUP02, ...
      const lastSupplier = await this.constructor.findOne({ code: /^SUP\d+$/ })
        .sort({ code: -1 })
        .select('code')
        .lean();
      
      if (!lastSupplier || !lastSupplier.code) {
        this.code = 'SUP01';
      } else {
        const lastNumber = parseInt(lastSupplier.code.replace('SUP', '')) || 0;
        const newNumber = String(lastNumber + 1).padStart(2, '0');
        this.code = `SUP${newNumber}`;
      }
    } catch (error) {
      console.error('Error generating supplier code:', error);
      this.code = `SUP${Date.now().toString().slice(-4)}`;
    }
  }
  next();
});

// NOTE: `StockTransaction` feature removed — no virtual totalTransactions

const Supplier = mongoose.model('Supplier', supplierSchema);

module.exports = Supplier;

File: frontend/lib/models/supplier.dart
class Supplier {
  final String id;
  final String code;
  final String name;
  final String? contactPerson;
  final String? phone;
  final String? email;
  final String? street;
  final String? district;
  final String? city;
  final String? country;
  final String? taxCode;
  final String? bankName;
  final String? accountNumber;
  final String? accountName;
  final bool isActive;
  final DateTime? createdAt;
  final String? createdByName;

  Supplier({
    required this.id,
    required this.code,
    required this.name,
    this.contactPerson,
    this.phone,
    this.email,
    this.street,
    this.district,
    this.city,
    this.country,
    this.taxCode,
    this.bankName,
    this.accountNumber,
    this.accountName,
    this.isActive = true,
    this.createdAt,
    this.createdByName,
  });

  String get fullAddress {
    final parts = <String>[];
    if (street != null && street!.isNotEmpty) parts.add(street!);
    if (district != null && district!.isNotEmpty) parts.add(district!);
    if (city != null && city!.isNotEmpty) parts.add(city!);
    if (country != null && country!.isNotEmpty) parts.add(country!);
    return parts.isNotEmpty ? parts.join(', ') : '-';
  }

  factory Supplier.fromJson(Map<String, dynamic> json) {
    final address = json['address'] as Map<String, dynamic>?;
    final bankAccount = json['bankAccount'] as Map<String, dynamic>?;

    return Supplier(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      code: json['code'] ?? '',
      name: json['name'] ?? '',
      contactPerson: json['contactPerson'] as String?,
      phone: json['phone'] as String?,
      email: json['email'] as String?,
      street: address?['street'] as String?,
      district: address?['district'] as String?,
      city: address?['city'] as String?,
      country: address?['country'] as String?,
      taxCode: json['taxCode'] as String?,
      bankName: bankAccount?['bankName'] as String?,
      accountNumber: bankAccount?['accountNumber'] as String?,
      accountName: bankAccount?['accountName'] as String?,
      isActive: json['isActive'] ?? true,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
      // transactionCount removed — backend returns 0 now
      createdByName: json['createdBy']?['fullName'] ?? json['createdBy']?['name'] ?? json['createdBy']?['username'],
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {
      'name': name,
      if (contactPerson != null && contactPerson!.isNotEmpty) 'contactPerson': contactPerson,
      if (phone != null && phone!.isNotEmpty) 'phone': phone,
      if (email != null && email!.isNotEmpty) 'email': email,
      if (taxCode != null && taxCode!.isNotEmpty) 'taxCode': taxCode,
    };

    // Address object
    if (street != null || district != null || city != null || country != null) {
      data['address'] = {
        if (street != null && street!.isNotEmpty) 'street': street,
        if (district != null && district!.isNotEmpty) 'district': district,
        if (city != null && city!.isNotEmpty) 'city': city,
        if (country != null && country!.isNotEmpty) 'country': country,
      };
    }

    // Bank account object
    if (bankName != null || accountNumber != null || accountName != null) {
      data['bankAccount'] = {
        if (bankName != null && bankName!.isNotEmpty) 'bankName': bankName,
        if (accountNumber != null && accountNumber!.isNotEmpty) 'accountNumber': accountNumber,
        if (accountName != null && accountName!.isNotEmpty) 'accountName': accountName,
      };
    }

    return data;
  }
}
